controllers/component_image_controller.go

```diff
@@ -142,150 +164,354 @@ func (r *ComponentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
+func (r *ComponentReconciler) ProvisionImageRepository(ctx context.Context, component *appstudioapiv1alpha1.Component) (bool, error) {
+	componentKey := types.NamespacedName{Namespace: component.Namespace, Name: component.Name}
+	log := r.Log.WithValues("ImageRepoProvisonForComponent", componentKey)
+	var err error
+
+	// Get or create repository provision status
+	imageRepoName := generateImageRepositoryName(component)
+	imageURL := fmt.Sprintf("quay.io/%s/%s", r.QuayOrganization, imageRepoName)
+	rps, exists := r.ImageRepositoryProvision[imageRepoName]
+	if !exists {
+		rps = NewImageRepositoryProvisionStatus()
+		r.ImageRepositoryProvision[imageRepoName] = rps
+	}
+
+	if !rps.isImageRepositoryCreated {
+		repo, err := r.QuayClient.CreateRepository(quay.RepositoryRequest{
+			Namespace:   r.QuayOrganization,
+			Visibility:  "public",
+			Description: "AppStudio repository for the user",
+			Repository:  imageRepoName,
+		})
+		if err != nil {
+			r.Log.Error(err, fmt.Sprintf("failed to create image repository %s", imageRepoName))
+			return false, err
+		}
+		if repo == nil {
+			return false, fmt.Errorf("unknown error in the image repository creation process")
 		}
-	} else if !errors.IsNotFound(err) {
-		log.Error(err, fmt.Sprintf("failed to read robot account secret %v", robotAccountSecretKey))
-		return ctrl.Result{}, err
+		log.Info(fmt.Sprintf("Image repository %s created for %v component", repo.Name, componentKey))
+
+		rps.isImageRepositoryCreated = true
 	}
 
-	if err := r.Client.Create(ctx, &robotAccountSecret); err != nil {
-		log.Error(err, fmt.Sprintf("error writing robot account token into Secret: %v", robotAccountSecretKey))
-		return ctrl.Result{}, err
+	var robotAccount *quay.RobotAccount
+	robotAccountName := generateRobotAccountName(component)
+
+	if !rps.isRobotAccountCreated {
+		robotAccount, err = r.QuayClient.CreateRobotAccount(r.QuayOrganization, robotAccountName)
+		if err != nil {
+			r.Log.Error(err, fmt.Sprintf("failed to create robot account %s", robotAccountName))
+			return false, err
+		}
+		if robotAccount == nil {
+			return false, fmt.Errorf("unknown error in the robot account creation process")
+		}
+		log.Info(fmt.Sprintf("Robot account %s created for %s image repository", robotAccount.Name, imageURL))
+
+		rps.isRobotAccountCreated = true
 	}
 
-	// Prepare data to update the component with
-	generatedRepository := RepositoryInfo{
-		Image:  imageURL,
-		Secret: robotAccountSecret.Name,
+	// Add finalizer to the component in order to clean up robot account and image repository if requested
+	if !controllerutil.ContainsFinalizer(component, ImageRepositoryFinalizer) && component.ObjectMeta.DeletionTimestamp.IsZero() {
+		if err := r.Client.Get(ctx, componentKey, component); err != nil {
+			log.Error(err, "failed to get component")
+			return false, err
+		}
+		if !controllerutil.ContainsFinalizer(component, ImageRepositoryFinalizer) && component.ObjectMeta.DeletionTimestamp.IsZero() {
```

Comment:
Why check this twice?

Reply:
Because component could get updates. We reread its configuration right above.